### PR TITLE
[RHOAIENG-6982] About Modal in the Dashboard

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Closes: [RHOAIENG-4220](https://issues.redhat.com/browse/RHOAIENG-4220)

## Description
Add an About dialog to the Dashboard UI using existing PatternFLy component.
https://www.patternfly.org/components/about-modal

This provides a place for UI users to see the current product version they are using, which can be useful for support and analytics use cases. 

## Screen shots
image

## How Has This Been Tested?
Tested locally

## Test Impact
Added e2e mock tests
Added jest test

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

/cc @simrandhaliw

### CODE REVIEW

**Comment by simrandhaliw** [comment]
@jeff-phillips-18 Could you provide a similar screenshot for the RHOAI About modal as well?

**Comment by jeff-phillips-18** [comment]
@simrandhaliw This is as close as I can get for now. The backend code that supplies some of the fields is not available on any server I know of. (I can install the image to a server if you need to see it full blown)

image

**Comment by simrandhaliw** [comment]
Thanks, @jeff-phillips-18. I assume that once all fields can render properly on a proper server, for RHOAI, we will display the correct logo and product name title.

However, I noticed that the version field labels for both ODH and RHOAI do not match the [designs](https://www.figma.com/proto/3qCdil1jk6QzlEEfs4TzRn/OpenShift-Console-About-modal?node-id=143-5264&t=Hs9ky6DakivWaMHK-0&scaling=scale-down&content-scaling=fixed&page-id=66%3A2&starting-point-node-id=143%3A5264). I will loop in @bredamc  and @kaedward to ensure they are aware.

For ODH, there has been an addition of 'Operator' in the label, whereas the designs specify 'Open Data Hub version'. I'm unsure if this addition is accurate.

Regarding RHOAI, the label does not display the product edition, i.e., 'Self-Managed' or 'Cloud Service'. So the version label should be either 'OpenShift AI Self-Managed version' or  'OpenShift AI Cloud Service version'.

File: backend/src/routes/api/odh-subscription/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/odh-subscription/index.ts:1]: Is there a reason we called it `odh-...`? Is this not for both?

File: frontend/src/app/App.scss [hunk 1]
```diff
@@ -78,3 +78,11 @@ body,
...
   }
 }
+
+// Override for about box height, reduce to fit the content instead of a fixed height
+.pf-v5-c-about-modal-box {
+  height: fit-content;
+  &__content {
+    grid-column-end: -1;
+  }
+}
```
Comment (andrewballantyne) [inline frontend/src/app/App.scss:88]: Hmmm -- is this necessary? This is to prevent a scrollbar? This seems like something the About Modal from PF should managed, no?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Initial, just code, feedback

File: backend/src/routes/api/odh-subscription/index.ts
Comment (jeff-phillips-18) [reply backend/src/routes/api/odh-subscription/index.ts:1]: Nope, just what came to mind. Suggestion?

File: frontend/src/app/App.scss (hunk 1 above)
Comment (jeff-phillips-18) [reply frontend/src/app/App.scss:88]: This is to allow the text to span more of the dialog. Without it the text and the info fields are limited in the space it can use. Here is a snapshot without this setting:

image

The PF team is assessing and will add a modifier to provide this update.

**Comment by jeff-phillips-18** [comment]
> However, I noticed that the version field labels for both ODH and RHOAI do not match the [designs](https://www.figma.com/proto/3qCdil1jk6QzlEEfs4TzRn/OpenShift-Console-About-modal?node-id=143-5264&t=Hs9ky6DakivWaMHK-0&scaling=scale-down&content-scaling=fixed&page-id=66%3A2&starting-point-node-id=143%3A5264). I will loop in @bredamc and @kaedward to ensure they are aware.
> 
> For ODH, there has been an addition of 'Operator' in the label, whereas the designs specify 'Open Data Hub version'. I'm unsure if this addition is accurate.
> 
> Regarding RHOAI, the label does not display the product edition, i.e., 'Self-Managed' or 'Cloud Service'. So the version label should be either 'OpenShift AI Self-Managed version' or 'OpenShift AI Cloud Service version'.

These values are not right here since they are retrieved from the backend (this is not yet available on any RHOAI server that I know of).  As for the `Operator` in the ODH version label. We can file an issue to have that changed but would not be part of this PR.

**Comment by simrandhaliw** [comment]
Other than what I mentioned earlier, everything else looks good to me.

File: backend/src/routes/api/odh-subscription/index.ts
Comment (andrewballantyne) [reply backend/src/routes/api/odh-subscription/index.ts:1]: `about-subscription`? I'm just against the `odh` prefix... you could call it `subscription` but it's not really getting you back _just a subscription_... so a bit misnomer imo.

**Comment by andrewballantyne** [comment]
> These values are not right here since they are retrieved from the backend

@jeff-phillips-18 @simrandhaliw the Operator team is aware of these needs right? Because we should be aligning with the product and the product should be aligning with the UX.

File: backend/src/routes/api/odh-subscription/index.ts
Comment (jeff-phillips-18) [reply backend/src/routes/api/odh-subscription/index.ts:1]: It is returning `SubscriptionStatusData` but it is the specific operator subscription so maybe  `operatorSubscriptionStatus` (though that is a bit wordy).

**Comment by jeff-phillips-18** [comment]
@simrandhaliw Can you point me to where the designs specify 'Open Data Hub version'.?

**Comment by simrandhaliw** [comment]
@jeff-phillips-18 Sure thing, [here is the design frame](https://www.figma.com/proto/3qCdil1jk6QzlEEfs4TzRn/OpenShift-Console-About-modal?node-id=143-5448&t=Hs9ky6DakivWaMHK-0&scaling=scale-down&content-scaling=fixed&page-id=66%3A2&starting-point-node-id=143%3A5264) where we display Open Data Hub version.

File: frontend/src/app/AboutDialog.scss [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// TODO: Remove when PF provides a fix for https://github.com/patternfly/patternfly/issues/6871
```
Comment (andrewballantyne) [inline frontend/src/app/AboutDialog.scss:1]: Logged https://issues.redhat.com/browse/RHOAIENG-9742 to track this on the flip side.

File: frontend/src/utilities/useWatchOperatorSubscriptionStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import * as React from 'react';
+import { fetchOperatorSubscriptionStatus } from '~/services/operatorSubscriptionStatusService';
+import { SubscriptionStatusData } from '~/types';
+import { useDeepCompareMemoize } from './useDeepCompareMemoize';
+import { POLL_INTERVAL } from './const';
+
+export const useWatchOperatorSubscriptionStatus = (): SubscriptionStatusData | null => {
```
Comment (andrewballantyne) [inline frontend/src/utilities/useWatchOperatorSubscriptionStatus.tsx:7]: out of curiosity, why didn't you use `useFetchState` -- it comes with a lot of the same functionality for near free

File: backend/src/routes/api/operator-subscription-status/index.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      const subscriptions = getSubscriptions();
+      const subNamePrefix = isRHOAI(fastify) ? 'rhods-operator' : 'opendatahub-operator';
+      const operatorSubscriptionStatus = subscriptions.find((sub) =>
+        sub.installedCSV?.includes(subNamePrefix),
+      );
+      if (operatorSubscriptionStatus) {
+        return operatorSubscriptionStatus;
+      }
+      fastify.log.error(`Failed to find operator subscription, ${subNamePrefix}`);
+      reply.send('Not found');
```
Comment (christianvogt) [inline backend/src/routes/api/operator-subscription-status/index.ts:19]: Shouldn't you return an error instead of a string?

eg. using `http-errors` lib:
```suggestion
      reply.send(reply.send(httpErrors.NotFound););
```

File: frontend/src/app/AboutDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+const RhoaiAboutText = `Red Hat® OpenShift® AI (formerly Red Hat OpenShift Data Science) is a flexible, scalable MLOps platform for data scientists and developers of artificial intelligence and machine learning (AI/ML) applications. Built using open source technologies, OpenShift AI supports the full lifecycle of AI/ML experiments and models, on premise and in the public cloud.`;
+const RhoaiDefaultReleaseName = `OpenShift AI`;
+
+const OdhAboutText = `Open Data Hub is an open source AI platform designed for the hybrid cloud. The community seeks to bridge the gap between application developers, data stewards, and data scientists by blending the leading open source AI tools with a unifying and intuitive user experience. Open Data Hub supports the full lifecycle of AI/ML experiments and models.`;
+const OdhDefaultReleaseName = `Open Data Hub operator`;
+
+interface AboutDialogProps {
+  onClose: () => void;
+  productName?: string; // Used only for testing
```
Comment (christianvogt) [inline frontend/src/app/AboutDialog.tsx:25]: Why not just mock `ODH_PRODUCT_NAME` ?

File: frontend/src/app/AboutDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        <h4>About</h4>
+        <p data-testid="about-text">{isRHOAI ? RhoaiAboutText : OdhAboutText}</p>
+      </TextContent>
+      <TextContent>
+        <div style={{ position: 'relative' }}>
+          {loading ? (
+            <Bullseye style={{ position: 'absolute', width: '100%' }}>
+              <Spinner size="xl" />
+            </Bullseye>
+          ) : null}
```
Comment (christianvogt) [inline frontend/src/app/AboutDialog.tsx:56]: If there's an error, we get a spinner indefinitely.

File: frontend/src/app/AboutDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+          <TextList component="dl" style={{ visibility: loading ? 'hidden' : 'visible' }}>
```
Comment (christianvogt) [inline frontend/src/app/AboutDialog.tsx:57]: Why a visibility style instead of part of the loading else clause for conditional rendering this section?
Is this done just to establish some white space for positioning the spinner mid dialog?

Comment (jeff-phillips-18) [reply frontend/src/app/AboutDialog.tsx:57]: Yes, it is to keep the dialog from changing size once the data is loaded.

**Comment by jeff-phillips-18** [comment]
> the Operator team is aware of these needs right? Because we should be aligning with the product and the product should be aligning with the UX.

@andrewballantyne  https://issues.redhat.com/browse/RHOAIENG-9751 tracks this issue

File: frontend/src/app/AboutDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import useFetchDsciStatus from '~/concepts/areas/useFetchDsciStatus';
+import { useWatchOperatorSubscriptionStatus } from '~/utilities/useWatchOperatorSubscriptionStatus';
+
+import './AboutDialog.scss';
+
+const RhoaiAboutText = `Red Hat® OpenShift® AI (formerly Red Hat OpenShift Data Science) is a flexible, scalable MLOps platform for data scientists and developers of artificial intelligence and machine learning (AI/ML) applications. Built using open source technologies, OpenShift AI supports the full lifecycle of AI/ML experiments and models, on premise and in the public cloud.`;
+const RhoaiDefaultReleaseName = `OpenShift AI`;
+
+const OdhAboutText = `Open Data Hub is an open source AI platform designed for the hybrid cloud. The community seeks to bridge the gap between application developers, data stewards, and data scientists by blending the leading open source AI tools with a unifying and intuitive user experience. Open Data Hub supports the full lifecycle of AI/ML experiments and models.`;
+const OdhDefaultReleaseName = `Open Data Hub`;
```
Comment (christianvogt) [inline frontend/src/app/AboutDialog.tsx:23]: @andrewballantyne are you ok with these hard coded values? Is it not possible to source these dynamically?

**Comment by christianvogt** [comment]
Works well.

I don't see the "edition" as requested in the final copy doc from UX: `<Product name (excluding “Red Hat”) with Edition> version`
These values are currently hard coded. Perhaps that's ok and there were some conversations around this already.

**Comment by jeff-phillips-18** [comment]
> Works well.
> 
> I don't see the "edition" as requested in the final copy doc from UX: `<Product name (excluding “Red Hat”) with Edition> version` These values are currently hard coded. Perhaps that's ok and there were some conversations around this already.

These values will be return the the RHOAI operator, I just don't have access to a server with the latest installed. See https://issues.redhat.com/browse/RHOAIENG-9326

**Comment by christianvogt** [comment]
Ah i see, the cluster i was testing on also didn't have `dsciStatus?.release?.name` set which is why it falls back to the hard coded values.

File: frontend/src/app/AboutDialog.tsx (hunk 8 above)
Comment (christianvogt) [reply frontend/src/app/AboutDialog.tsx:23]: The release names here are fallback values when `dsciStatus?.release?.name` is undefined.
